harness/determined/core/_profiler.py

```diff
@@ -0,0 +1,519 @@
+
+        if not isinstance(samples_per_report, int) or samples_per_report < 1:
+            raise ValueError(
+                f"Samples per report specifies the number of samples to aggregate before "
+                f"reporting the metric. It must be an int > 1, but was specified as "
+                f"{samples_per_report}."
+            )
+
+        # Currently, metrics collected are scoped at the machine level, so we only collect metrics
+        # on the chief worker of each node.
+        if self._distributed.local_rank != 0:
+            return
+
+        logger.info("Starting system metrics profiling.")
+
+        metrics_queue: queue.Queue = queue.Queue()
+        self._collector = _Collector(
+            metrics_queue=metrics_queue,
+            sampling_interval=sampling_interval,
+            aggregation_period=samples_per_report,
+        )
+        self._shipper = _Shipper(
+            session=self._session,
+            agent_id=self._agent_id,
+            trial_id=self._trial_id,
+            run_id=self._run_id,
+            metrics_queue=metrics_queue,
+        )
+
+        self._collector.start()
+        self._shipper.start()
+        self._on = True
+
+    def off(self) -> None:
+        """Turns off profiling.
+
+        Sets the internal state of this class and stops any threads that where created.
+
+        .. note::
+
+            This method is idempotent: if profiling is already off, this method is effectively a
+            no-op.
+        """
+        if not self._on:
+            return
+
+        logger.info("Stopping system metrics profiling.")
+        self._close()
+        self._on = False
+
+    def _close(self) -> None:
+        """Shuts down any threads that were created."""
+        if self._collector:
+            self._collector.stop()
+        if self._shipper:
+            self._shipper.stop()
+
+
+class DummyProfilerContext(ProfilerContext):
+    """Drop-in replacement of ``ProfilerContext``.
+
+    Used by the ``core.Context`` for cases when profiling cannot run.
+    """
+
+    def __init__(self) -> None:
+        pass
+
+    def on(self, sampling_interval: int = 1, samples_per_report: int = 10) -> None:
+        pass
+
+    def off(self) -> None:
+        pass
+
+    def _close(self) -> None:
+        pass
+
+
+def _average_metric_samples_depth_one(metric_samples: List[Dict[str, Any]]) -> Dict[str, Any]:
+    """Helper method to merge a list of dictionary averaging their values by their keys.
+
+    Supports up to 1 level of nesting. Returns a single merged dictionary where the values are
+    averaged across all dictionaries in the given list by key.
+    # TODO (anda): find a cleaner way to do this.
+    """
+    aggregated_metrics: Dict[str, Any] = {}
+    for sample in metric_samples:
+        for k, v in sample.items():
+            if isinstance(v, dict):
+                aggregated_metrics[k] = aggregated_metrics.get(k, {})
```

Comment:
nit: little clearer with a `defaultdict`, I think.

Reply:
```python
aggregated_metrics = defaultdict(int)
aggregated_metrics = defaultdict(defaultdict(int))
```

Reply:
eh, don't love the way this reads. `.get(k, {})` is easily readable IMO. i do think `defaultdict` is best-practices way of doing this, but since i have a ticket to refactor this method anyway, i don't think it's worth it.